# Đồ án cuối kỳ:
(Cập nhật lần cuối: 04/07/2021)\
Nhóm 7
Thành viên:
 1. 18120062 - Nguyễn Hoàng Nam
 2. 18120096 - Trương Đại Triều

----

<h1>Bảng nội dung<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Import-thư-viện" data-toc-modified-id="Import-thư-viện-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Import thư viện</a></span></li><li><span><a href="#Thu-thập-dữ-liệu" data-toc-modified-id="Thu-thập-dữ-liệu-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Thu thập dữ liệu</a></span></li><li><span><a href="#Khám-phá-dữ-liệu-&-tiền-xử-lý-dữ-liệu" data-toc-modified-id="Khám-phá-dữ-liệu-&-tiền-xử-lý-dữ-liệu-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Khám phá dữ liệu & tiền xử lý dữ liệu</a></span></li><li><span><a href="#Phân-tích-để-trả-lời" data-toc-modified-id="Phân-tích-để-trả-lời-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Phân tích để trả lời</a></span></li><li><span><a href="#Nhìn-lại-quá-trình-làm-đồ-án" data-toc-modified-id="Nhìn-lại-quá-trình-làm-đồ-án-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Nhìn lại quá trình làm đồ án</a></span></li><li><span><a href="#Tài-liệu-tham-khảo" data-toc-modified-id="Tài-liệu-tham-khảo-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Tài liệu tham khảo</a></span></li></ul></div>

---

# Import thư viện

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from datetime import datetime

---

# Thu thập dữ liệu

## Dữ liệu của bạn là về chủ đề gì và bạn lấy từ nguồn nào?

Dữ liệu nói về việc tiêm vaccine phòng Covid-19, ở nhiều quốc gia được cập nhập thường xuyên. Dữ liệu được lưu tại 

(https://github.com/owid/covid-19-data/tree/master/public/data/vaccinations)

##  Người ta có cho phép bạn dùng dữ liệu như này không?


License gốc được public trên trang [OurWorldInData](https://ourworldindata.org/coronavirus), các dataset được cấp quyền tái sử dụng tự do "[Reuse our work freely](https://ourworldindata.org/coronavirus#licence)". Đây là trang web cung cấp đầy đủ thông tin, về tình hình dịch bệnh, số ca nhiễm, tử vong, phân phối vacine trên từng quốc gia, vùng lãnh thổ. Trang web được phát triển bởi trường [Đại học Oxford, Anh](https://www.oxfordmartin.ox.ac.uk/global-development) và tổ chức phi lợi nhuận [Global Change Data Lab](https://global-change-data-lab.org/) 


##  Người ta thu thập dữ liệu như thế nào?

Dữ liệu được các tổ chức tại các quốc gia cập nhập hiện tại có đã bao gồm dữ liệu của 169 quốc gia.

---

# Khám phá dữ liệu & tiền xử lý dữ liệu

## Đọc dữ liệu từ file csv vào dataframe

In [3]:
vaccinations=pd.read_csv('data.csv')
vaccinations.head()

location iso_code        date  total_vaccinations  people_vaccinated  \
0  Afghanistan      AFG  2021-02-22                 0.0                0.0   
1  Afghanistan      AFG  2021-02-23                 NaN                NaN   
2  Afghanistan      AFG  2021-02-24                 NaN                NaN   
3  Afghanistan      AFG  2021-02-25                 NaN                NaN   
4  Afghanistan      AFG  2021-02-26                 NaN                NaN   

   people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
0                      NaN                     NaN                 NaN   
1                      NaN                     NaN              1367.0   
2                      NaN                     NaN              1367.0   
3                      NaN                     NaN              1367.0   
4                      NaN                     NaN              1367.0   

   total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
0                             0.0                            0.0   
1                             NaN                            NaN   
2                             NaN                            NaN   
3                             NaN                            NaN   
4                             NaN                            NaN   

   people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  
0                                  NaN                             NaN  
1                                  NaN                            35.0  
2                                  NaN                            35.0  
3                                  NaN                            35.0  
4                                  NaN                            35.0

## Dữ liệu có bao nhiêu dòng và bao nhiêu cột?

Dữ liệu có tổng cộng **30809** dòng và **12** cột

In [4]:
shape = vaccinations.shape
n_rows=shape[0]
n_cols=shape[1]
print(n_rows)
print(n_cols)

30809
12


## Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Mỗi dòng cho biết số người đã tiêm vaccine, loại vaccine, đã tiêm đủ 2 mũi chưa cho trong các ngày được ghi nhận. Ngoài ra cũng có biết thêm thông tin số người bị bệnh theo tỉ lệ để thể hiện được quy mô của dịch bênh theo quốc gia, khu vực. Do các dòng đều được các tổ chức lớn thêm nên ít có dòng sai khác.

## Dữ liệu có các dòng bị lặp không?

In [5]:
vaccinations[vaccinations.duplicated()].shape[0] != 0

False

## Mỗi cột có ý nghĩa gì?¶

**Ý nghĩa của từng cột:**
- **location**: Tên quốc gia (hoặc vùng lãnh thổ trong 1 quốc gia(.
- **iso_code**: ISO 3166-1  và 3 ký tự cuối là mã quốc gia.
- **date**: Ngày quan trắc.
- **total_vaccinations**: Tổng số liều đã được sử dụng. Đơn vị tính là liều đã tiêm, có thể không bằng tổng số người đã tiêm. Nếu một người tiêm một liều thì đếm tăng lên 1 đơn vị. Nếu họ tiêm liều thứ 2 thì cũng tăng thêm 1 đơn vị
- **people_vaccinated**: Tổng số người nhận ít nhất một mũi tiêm. Nếu một người đã tiêm 1 mũi, thì đếm tăng lên 1 đơn vị. Nếu họ được tiêm liều thứ 2, thi giữ nguyên.
- **people_fully_vaccinated**: Tổng số người đã tiêm đủ 2 liều vaccine. Nếu một người nhận đủ 2 liều vaccine, thì đếm tăng lên 1. Nếu học chỉ mới tiêm 1 liều thì giữ nguyên
- **daily_vaccinations_raw**: Số lượng liều vaccine được sử dụng mỗi ngày, cập nhật theo ngày.
- **daily_vaccintions**:Số lượng liều vaccine được sử dụng mỗi ngày, cập nhật theo ngày. Dữ liệu đã được chuẩn hóa hơn.
- **total_vaccinations_per_hundred**: total_vaccinations trên 100 dân trong tổng dân số của quôc gia đó.
- **people_vaccinated_per_hundred**: people_vaccinated trên 100 dân trong tổng dân số quốc gia đó.
- **people_fully_vaccinated_per_hundred**: Tỗng số người đã được tiểm đủ 2 liều vaccine. Nếu một người tiêm liều đầu tiên thì giữ nguyên. Nếu họ nhận liều thứ 2, số liệu tính tăng thêm 1 đơn vị.
- **daily_vaccinations_per_million**: daily_vaccintions trên 1,000,000 dân trong tổng dân số của quốc gia đó.

**Note:**\
    - Mô tả trên được dịch từ bản [mô tả gốc](https://github.com/owid/covid-19-data/blob/master/public/data/vaccinations/README.md)

## Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

In [6]:
vaccinations.dtypes

location                                object
iso_code                                object
date                                    object
total_vaccinations                     float64
people_vaccinated                      float64
people_fully_vaccinated                float64
daily_vaccinations_raw                 float64
daily_vaccinations                     float64
total_vaccinations_per_hundred         float64
people_vaccinated_per_hundred          float64
people_fully_vaccinated_per_hundred    float64
daily_vaccinations_per_million         float64
dtype: object

**Nhận xét:**
1. Cần chuyển cột date về kiểu dữ liệu dạng date

## Số-lượng/tỉ-lệ các giá trị thiếu?

In [7]:
number_of_null = vaccinations.isna().sum()
ratio_of_null = vaccinations.isnull().sum() * 100 / len(vaccinations)

number_ratio_of_null = pd.concat([number_of_null, ratio_of_null], axis=1)
number_ratio_of_null.columns = ["Number of null", "Ratio of null"]

number_ratio_of_null

Number of null  Ratio of null
location                                          0       0.000000
iso_code                                          0       0.000000
date                                              0       0.000000
total_vaccinations                            12579      40.828979
people_vaccinated                             13403      43.503522
people_fully_vaccinated                       16378      53.159791
daily_vaccinations_raw                        15388      49.946444
daily_vaccinations                              288       0.934792
total_vaccinations_per_hundred                12579      40.828979
people_vaccinated_per_hundred                 13403      43.503522
people_fully_vaccinated_per_hundred           16378      53.159791
daily_vaccinations_per_million                  288       0.934792

## Tiền xử lý bỏ cột có giá trị thiếu nhiều và bỏ cột chúng tôi không hiểu rõ ý nghĩa

**Nhận xét:**
1. Cần chuyển cột date về kiểu dữ liệu dạng date
2. Cột daily_vaccinations_raw có rất nhiều giá trị null, dữ liệu chưa được chuẩn hóa. Có thể thay thế bằng cột daily_vaccinations đã được tác giả chuẩn hóa sẵn
3. Các cột people_fully_vaccinated, people_fully_vaccinated_per_hundred dữ liệu không rõ ràng, cần lược bỏ
4. Các cột people_fully_vaccinated, total_vaccinations , total_vaccinations_per_hundred, people_vaccinated_per_hundred, daily_vaccinations có giá trị null nhiều cần điền đầy đủ 
5. Dữ liệu này không chỉ bao gồm các quốc mà còn thống kê theo vùng lãnh thổ, theo thu nhập. Nên chia dataset vaccinations ban đầu thành 3 phần, 1 của quốc gia cụ thể, 2 của vùng lãnh thổ, 3 theo thu nhập

### Điều chỉnh kiểu dữ liệu cho phù hợp cho các cột

### 1. Chuyển dtype cột 'date' sang datetime

In [6]:
vaccinations['date']=pd.to_datetime(vaccinations['date'],format="%m/%d/%Y")

### 2. Điền vào giá trị null của các cột
Thực hiện với 6 cột sau:
- Điền giá trị null dựa vào giá trị của ngày trước đó cùng cột:
    1. people_vaccinated
    2. total_vaccinations 
    3. total_vaccinations_per_hundred
    4. people_vaccinated_per_hundred
- Điền giá trị null dựa vào giá trị của ngày sau đó cùng cột
    5. daily_vaccinations
    6. daily_vaccinations_per_million

In [7]:
vaccinations['total_vaccinations'].fillna(method='ffill', inplace =True)
vaccinations['people_vaccinated'].fillna(method='ffill', inplace =True)
vaccinations['total_vaccinations_per_hundred'].fillna(method='ffill', inplace =True)
vaccinations['people_vaccinated_per_hundred'].fillna(method='ffill', inplace =True)
vaccinations['daily_vaccinations'].fillna(method='bfill', inplace =True)
vaccinations['daily_vaccinations_per_million'].fillna(method='bfill', inplace =True)

### 3. Loại bỏ các cột không rõ nghĩa, thiếu nhiều:
Thực hiện với 3 cột sau:
1. people_fully_vaccinated
2. people_fully_vaccinated_per_hundred 
3. daily_vaccinations_raw

In [8]:
vaccinations.drop(columns=['people_fully_vaccinated','people_fully_vaccinated_per_hundred','daily_vaccinations_raw'],inplace =True)


### 4. Chia dataset thành 3 thành phần:
1. Vùng lãnh thổ
2. Theo thu nhập
3. Quốc gia cụ thể
**Nhận xét:**\
    - Mã iso_code của vùng lãnh thổ thường bắt đầu bằng OWID
    - Có vài quốc gia thuộc một liên minh hay tình trạng bất ổn được thống kê 2 lần cần loại bỏ

#### 4.1. Vùng lãnh thổ

In [10]:
# lấy các hàng dữ liệu thuộc về vùng lãnh thổ
boolResult = vaccinations.apply(lambda x: 'OWID' in x['iso_code'],axis=1)
vacc = vaccinations[boolResult]

diffs,counts=np.unique(vacc['location'],return_counts=True)
diffs

array(['Africa', 'Asia', 'England', 'Europe', 'European Union',
       'High income', 'Kosovo', 'Low income', 'Lower middle income',
       'North America', 'Northern Cyprus', 'Northern Ireland', 'Oceania',
       'Scotland', 'South America', 'Upper middle income', 'Wales',
       'World'], dtype=object)

**Nhận xét**
1. Các vùng lãnh thổ được liệt kê là: Africa, European Union, North America, Oceania, South America, World

In [11]:
regions=np.array(['Africa', 'Asia', 'Europe', 'Oceania', 'South America', 'North America', 'World'])
boolResult= vacc.apply(lambda x: x['location'] in regions, axis=1)
regional_vacc= vacc[boolResult]
regional_vacc.sample(5)

location  iso_code       date  total_vaccinations  people_vaccinated  \
8803   Europe  OWID_EUR 2021-02-08          33756595.0         27321314.0   
8859   Europe  OWID_EUR 2021-04-05         138991335.0        102134963.0   
186    Africa  OWID_AFR 2021-03-07           5053898.0          4450336.0   
8900   Europe  OWID_EUR 2021-05-16         294471551.0        202874096.0   
187    Africa  OWID_AFR 2021-03-08           5402174.0          4724256.0   

      daily_vaccinations  total_vaccinations_per_hundred  \
8803           1103295.0                            4.51   
8859           2254595.0                           18.56   
186             168558.0                            0.38   
8900           4450157.0                           39.33   
187             187189.0                            0.40   

      people_vaccinated_per_hundred  daily_vaccinations_per_million  
8803                           3.65                          1474.0  
8859                          13.64                          3011.0  
186                            0.33                           126.0  
8900                          27.10                          5944.0  
187                            0.35                           140.0

#### 4.2. Theo thu nhập


In [12]:
incomes=np.array(['High income', 'Low income', 'Lower middle income', 'Upper middle income'])
boolResult= vacc.apply(lambda x: x['location'] in incomes, axis=1)
income_vacc= vacc[boolResult]
income_vacc.sample(5)

location  iso_code       date  total_vaccinations  \
11974          High income  OWID_HIC 2021-04-04        3.373198e+08   
11904          High income  OWID_HIC 2021-01-24        4.633090e+07   
29539  Upper middle income  OWID_UMC 2021-06-29        1.623704e+09   
29399  Upper middle income  OWID_UMC 2021-02-09        5.118797e+07   
29401  Upper middle income  OWID_UMC 2021-02-11        5.242991e+07   

       people_vaccinated  daily_vaccinations  total_vaccinations_per_hundred  \
11974        225399991.0           6091076.0                           26.71   
11904         38155839.0           2622787.0                            3.67   
29539        894537804.0          27731690.0                           61.16   
29399          7363815.0           2760078.0                            1.93   
29401          8245654.0           1718783.0                            1.97   

       people_vaccinated_per_hundred  daily_vaccinations_per_million  
11974                          17.85                          4822.0  
11904                           3.02                          2076.0  
29539                          33.69                         10446.0  
29399                           0.28                          1040.0  
29401                           0.31                           647.0

#### 4.3. Quốc gia cụ thể

In [13]:
# lấy các hàng dữ liệu thuộc về quốc gia
boolResult = vaccinations.apply(lambda x: 'OWID' not in x['iso_code'],axis=1)
national_vacc = vaccinations[boolResult]
national_vacc.sample(5)

location iso_code       date  total_vaccinations  people_vaccinated  \
21223         Oman      OMN 2021-01-26             36206.0            24773.0   
21595    Palestine      PSE 2021-06-04            607122.0           372827.0   
1815     Australia      AUS 2021-05-09           2654338.0           162551.0   
9778        France      FRA 2021-04-12          15115069.0         11284068.0   
19860  New Zealand      NZL 2021-06-24           1069552.0           662609.0   

       daily_vaccinations  total_vaccinations_per_hundred  \
21223              1165.0                            0.71   
21595             13150.0                           11.90   
1815              57181.0                           10.41   
9778             343531.0                           22.37   
19860             18453.0                           22.18   

       people_vaccinated_per_hundred  daily_vaccinations_per_million  
21223                           0.49                           228.0  
21595                           7.31                          2578.0  
1815                            0.64                          2242.0  
9778                           16.70                          5085.0  
19860                          13.74                          3827.0

### Xem lại dữ liệu sau khi đã điều chỉnh

In [14]:
vaccinations.dtypes

location                                  object
iso_code                                  object
date                              datetime64[ns]
total_vaccinations                       float64
people_vaccinated                        float64
daily_vaccinations                       float64
total_vaccinations_per_hundred           float64
people_vaccinated_per_hundred            float64
daily_vaccinations_per_million           float64
dtype: object

## Với mỗi cột có kiểu dữ liệu dạng số (numeric), các giá trị được phân bố như thế nào?

In [17]:
national_vacc[['date',]].agg([min,max])

date
min 2020-12-02
max 2021-07-03

Vì các dòng dữ liệu là thống kê chương trình vaccine của từng quốc gia, khu vực, dược quan trắc từng ngày từ ngày triển khai đến thời điểm cập nhật gần nhất. Nên ta chọn 1 mốc là ngày **01/06/2021** để xem phân bố của các giá trị (ở mốc này đầy đủ nhất các quốc gia cung cấp và cập nhật số liệu) 

In [18]:
cur_date= pd.to_datetime('2021/06/01', format='%Y/%m/%d')
temp= national_vacc[national_vacc['date']==cur_date]
temp[['total_vaccinations','people_vaccinated']].agg([max,min,'mean','sum'])


total_vaccinations  people_vaccinated
max         6.819080e+08       1.695852e+08
min         0.000000e+00       0.000000e+00
mean        9.828503e+06       4.311474e+06
sum         1.985358e+09       8.709177e+08

Nhìn vào số liệu trên, đến ngày 01/06/2021 đã có gần 2 tỷ liều vaccine đã được sử dụng. Số người được tiêm ít nhất một mũi là 871 triệu người. Số lượng vaccine nhiều nhất mà 1 quốc gia đã sử dụng là 682 triệu liều. Và  số người tiêm chủng ít nhất một liều nhiều nhất của 1 quốc gia là gần 170 triệu

In [20]:
cur_date= pd.to_datetime('2021/06/01', format='%Y/%m/%d')
temp= national_vacc[national_vacc['date']==cur_date]
temp[['total_vaccinations_per_hundred','people_vaccinated_per_hundred','daily_vaccinations','daily_vaccinations_per_million']].agg([min,max,'mean'])


total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
min                         0.000000                       0.000000   
max                       230.090000                     116.000000   
mean                       33.363465                      21.224703   

      daily_vaccinations  daily_vaccinations_per_million  
min         0.000000e+00                        0.000000  
max         1.931343e+07                    31940.000000  
mean        1.692522e+05                     3923.940594

## Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical), các giá trị được phân bố như thế nào?

Có 2 cột dữ liệu dạng phân loại là **location**, **iso_code**. Số lượng dòng dữ liệu từng quốc gia cho biết số ngày quốc gia đó đã triển khai chương trinh tiêm chủng quốc gia.

In [27]:
diffs,counts=np.unique(national_vacc['location'],return_counts=True)
print(len(diffs))

211


Vaccine đã tiếp cận được với 211 quốc gia

---


# Xác định các câu hỏi

## Câu hỏi là gì?

## Nếu trả lời được câu hỏi thì sẽ có lợi ích gì?

---


# Phân tích để trả lời

---


# Nhìn lại quá trình làm đồ án

---


# Tài liệu tham khảo